homeassistant/components/sensor/citybikes.py

```diff
@@ -0,0 +1,310 @@
+    vol.Required(ATTR_EMPTY_SLOTS): cv.positive_int,
+    vol.Required(ATTR_LATITUDE): vol.Coerce(float),
+    vol.Required(ATTR_LONGITUDE): vol.Coerce(float),
+    vol.Required(ATTR_ID): cv.string,
+    vol.Required(ATTR_NAME): cv.string,
+    vol.Required(ATTR_TIMESTAMP): cv.string,
+    vol.Optional(ATTR_EXTRA): vol.Schema({
+        vol.Optional(ATTR_UID): cv.string
+        }, extra=vol.REMOVE_EXTRA)
+    }, extra=vol.REMOVE_EXTRA)
+
+STATIONS_RESPONSE_SCHEMA = vol.Schema({
+    vol.Required(ATTR_NETWORK): vol.Schema({
+        vol.Required(ATTR_STATIONS_LIST): [STATION_SCHEMA]
+        }, extra=vol.REMOVE_EXTRA)
+    })
+
+
+MONITORED_NETWORKS = {}
+
+
+# pylint: disable=unused-argument
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_entities,
+                         discovery_info=None):
+    """Set up the CityBikes platform."""
+    @asyncio.coroutine
+    def async_setup_network(now=None):
+        """Set up a network with stations without blocking."""
+        latitude = config.get(CONF_LATITUDE, hass.config.latitude)
+        longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+        network_id = config.get(CONF_NETWORK)
+        stations_list = set(config.get(CONF_STATIONS_LIST, []))
+        radius = config.get(CONF_RADIUS, 0)
+        name = config.get(CONF_NAME)
+        if not hass.config.units.is_metric:
+            radius = distance.convert(radius, LENGTH_FEET, LENGTH_METERS)
+
+        if not network_id:
+            network_id = yield from CityBikesNetwork.get_closest_network_id(
+                hass, latitude, longitude)
+            if not network_id:
+                async_track_point_in_utc_time(hass, async_setup_network,
+                                              dt_util.utcnow() + SCAN_INTERVAL)
+                return
+
+        if network_id not in MONITORED_NETWORKS:
+            network = CityBikesNetwork(hass, network_id)
+            MONITORED_NETWORKS[network_id] = network
+            hass.async_add_job(network.async_refresh)
+            async_track_time_interval(hass, network.async_refresh,
+                                      SCAN_INTERVAL)
+        else:
+            network = MONITORED_NETWORKS[network_id]
+
+        while not network.ready:
+            yield from asyncio.sleep(0.5)
+
+        entities = []
+        for station in network.stations:
+            dist = location.distance(latitude, longitude,
+                                     station[ATTR_LATITUDE],
+                                     station[ATTR_LONGITUDE])
+            station_id = station[ATTR_ID]
+            station_uid = str(station.get(ATTR_EXTRA, {}).get(ATTR_UID, ''))
+
+            if radius > dist or stations_list.intersection((station_id,
+                                                            station_uid)):
+                entities.append(CityBikesStation(network, station_id, name))
+
+        async_add_entities(entities, True)
+
+    hass.async_add_job(async_setup_network)
```

Comment:
Since the setup requires communication with the remote API, I want it to be a non-blocking task - otherwise, the start up of Home Assistant is blocked until this is set up.

Furthermore, if network connectivity is unavailable on start-up (for instance, after a power outage, the Home Assistant instance might boot up before the Internet router), I'd like the platform to perform additional attempts at setting itself up, by scheduling the setup routine on failure. If I'm doing that, I thought I might as well start with the scheduling from the beginning, and avoid blocking while I'm at it.